File: Source/Fantasy.Jobs/Resources/WaitForResourceProvider.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Fantasy.Jobs.Management;
using System.Runtime.Caching;
using System.Security.Cryptography;
using System.IO;
using Fantasy.Jobs.Utils;

namespace Fantasy.Jobs.Resources
{
    public class WaitForResourceProvider : ObjectWithSite, IResourceProvider
    {

        private MemoryCache _cache;

        private object _syncRoot = new object();

        private static readonly CacheItemPolicy _cachePolicy = new CacheItemPolicy() { SlidingExpiration = new TimeSpan(1, 0, 0) };

        public bool CanHandle(string name)
        {
            return string.Equals(name, "WaitFor", StringComparison.OrdinalIgnoreCase);
        }

        public void Initialize()
        {
            _cache = new MemoryCache("WaitForResourceProvider");
        }

        private string GetKey(string ids)
        {
            MemoryStream stream = new MemoryStream(Encoding.Unicode.GetBytes(ids));
            return MD5HashCodeHelper.GetMD5HashCode(stream);
        }

        public bool IsAvailable(ResourceParameter parameter)
        {

            string ids = parameter.Values.GetValueOrDefault("jobs", string.Empty);
            if (!String.IsNullOrEmpty(ids))
            {
                WaitForMode model = (WaitForMode)Enum.Parse(typeof(WaitForMode), parameter.Values.GetValueOrDefault("mode", "All"), true);
                string key = model.ToString() + this.GetKey(ids);
                List<Guid> waitList;
                lock (_syncRoot)
                {
                    waitList = (List<Guid>)this._cache[key];
                    if (waitList == null)
                    {
                        waitList = (from id in parameter.Values.GetValueOrDefault("jobs", string.Empty).Split(new char[] { ';' }, StringSplitOptions.RemoveEmptyEntries) select new Guid(id)).ToList();
                    }

                    _cache.Add(key, waitList, _cachePolicy);
                }

                if (waitList.Count > 0)
                {
                    lock (waitList)
                    {
                        List<Guid> temp = new List<Guid>(waitList);
                        IJobQueue queue = this.Site.GetRequiredService<IJobQueue>();

                        if (model == WaitForMode.All)
                        {
                            foreach (Guid id in temp)
                            {
                                JobMetaData job = queue.FindJobMetaDataById(id);
                                if (! queue.IsTerminated(id))
                                {
                                    return false;
                                }
                                else
                                {
                                    waitList.Remove(id);
                                }
                            }
                            return true;
                        }
                        else
                        {
                            foreach (Guid id in temp)
                            {
                                if (queue.IsTerminated(id))
                                {
                                    waitList.Clear();
                                    return true;
                                }
                            }
                            return false;
                        }
                    }
                }
                else
                {
                    return true;
                }

            }
            else
            {
                return true;
            }
            
        }

        public bool Request(ResourceParameter parameter, out object resource)
        {
            resource = null;
            return this.IsAvailable(parameter);
        }

        public void Release(object resource)
        {
            
        }

        public event EventHandler Available {add{}remove {}}

        public event EventHandler<ProviderRevokeArgs> Revoke {add{} remove {}}
    }
}

File: Fantasy.Jobs/src/fantasy/jobs/resources/WaitForResourceProvider.java
package fantasy.jobs.resources;

import java.util.*;

import org.apache.commons.lang3.StringUtils;

import fantasy.*;
import fantasy.collections.*;

import fantasy.jobs.management.*;

import fantasy.runtime.caching.SimpleCache;

public class WaitForResourceProvider extends ResourceProvider implements IResourceProvider
{

	private SimpleCache<ArrayList<UUID>> _cache = new SimpleCache<ArrayList<UUID>>(60*60, 60*60);

	private Object _syncRoot = new Object();

	

	public final boolean canHandle(String name)
	{
		return StringUtils.equalsIgnoreCase("WaitFor", name);
		
	}

	public final void initialize()
	{
		
	}

	private String getKey(String ids) throws Exception
	{
	
		return MD5HashCodeHelper.compute(ids);
	}

	public final boolean isAvailable(ResourceParameter parameter) throws Exception
	{

		String ids = parameter.getValues().get("jobs");
		if (!StringUtils2.isNullOrEmpty(ids))
		{

			WaitForMode model = WaitForMode.valueOf(MapUtils.getValueOrDefault(parameter.getValues(), "mode", "All"));
			String key = model.toString() + this.getKey(ids);
			java.util.ArrayList<UUID> waitList;
			synchronized (_syncRoot)
			{
				waitList = this._cache.get(key);
				if (waitList == null)
				{
					waitList = new Enumerable<String>(StringUtils2.split(ids, ";",true)).select(new Selector<String, UUID>(){

						@Override
						public UUID select(String item) {
							
							return UUID.fromString(item);
						}}).toArrayList();  
				}

				_cache.put(key, waitList);
			}

			if (waitList.size() > 0)
			{
				synchronized (waitList)
				{
					java.util.ArrayList<UUID> temp = new java.util.ArrayList<UUID>(waitList);
					IJobQueue queue = this.getSite().getRequiredService(IJobQueue.class);

					if (model == WaitForMode.All)
					{
						for (UUID id : temp)
						{
							
							if (!queue.IsTerminated(id))
							{
								return false;
							}
							else
							{
								waitList.remove(id);
							}
						}
						return true;
					}
					else
					{
						for (UUID id : temp)
						{
							if (queue.IsTerminated(id))
							{
								waitList.clear();
								return true;
							}
						}
						return false;
					}
				}
			}
			else
			{
				return true;
			}

		}
		else
		{
			return true;
		}

	}

	public final boolean request(ResourceParameter parameter, RefObject<Object> resource) throws Exception
	{
		resource.argvalue = null;
		return this.isAvailable(parameter);
	}

	public final void release(Object resource)
	{

	}
}